### PR TITLE
Grammar: Adds support for materialized views for postgres dialect

### PR DESCRIPTION
**Pull request by derickl**
### Brief summary of the change made
Partially addresses #2039. This PR adds support for materialized views for the postgres dialect.
This is my first attempt working through the SQLfluff codebase, so there could be a lot missing.

What's addressed:
- CREATE MATERIALIZED VIEW 
- ALTER MATERIALIZED VIEW 
- DROP MATERIALIZED VIEW 
- REFRESH MATERIALIZED VIEW

### Are there any other side effects of this change that we should be aware of?
None. 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by derickl** [comment]
`CREATE MATERIALIZED VIEW` supports a `WITH [NO] DATA`  qualifier similar to `CREATE TABLE [name] AS`. I wasn't able to get the query below to parse without enclosing the `SELECT` block in parenthesis. I noticed the same behaviour in [CREATE TABLE [name] AS](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_postgres.py#L1057) and it's [test queries](https://github.com/sqlfluff/sqlfluff/blob/main/test/fixtures/dialects/postgres/postgresl_create_table_as.sql#L46-L58)

**Fails**
```sql
CREATE MATERIALIZED VIEW bar
AS
SELECT
  col
FROM
  table
WITH DATA;
```

**Parses**
```sql
CREATE MATERIALIZED VIEW bar
AS
(
  SELECT
    col
  FROM
    table
)
WITH DATA;
```

**Comment by jpy-git** [comment]
The `WITH [NO] DATA` brackets vs no brackets one is interesting. I imagine without the brackets the parser doesn't know when the SELECT statement ends, so can't tell if `WITH DATA` is part of the CREATE or the SELECT. If you have access to postgres it might be worth checking quickly to confirm the without brackets case is valid SQL? May well be, but good to check first 😄 

The only example I could find of `WITH [NO] DATA` was
```SQL
CREATE TABLE odetails_copy_nodata AS TABLE odetails
WITH NO DATA;
```
(https://www.datasciencemadesimple.com/create-the-copy-of-table-in-postgresql-with-and-without-data/)

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1342]: Worth updating these doc links to postgres 14 and checking the grammar matches that so we're up to date 👍

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1390]: If we're not aiming to add the complete syntax, could you just add a TODO note to the docstring let people know that it is incomplete and which cases it covers.

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1433]:
```suggestion
        Delimited(Ref("TableReferenceSegment")),
```

This covers dropping multiple materialized views (`[, ...]` in the docs)
```SQL
DROP MATERIALIZED VIEW foo, bar;
```
(add the above to the unit tests as well)

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1410]:
```suggestion
                    "CURRENT_USER",
                    "SESSION_USER",
                    "CURRENT_ROLE",
```
Also need CURRENT_ROLE as this was added in Postgres 14

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1368]:
```suggestion
                        Sequence(
                            Ref("ParameterNameSegment"),
                            Sequence(
                                Ref("EqualsSegment"),
                                Ref("LiteralGrammar"),
                                optional=True,
                            ),
                        ),
```
looking in the docs the equals value part is optional: `storage_parameter [= value]`

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1369]: you don't need this line as it's comma delimited by default

**Review by jpy-git — COMMENTED** [review]
A couple points to address, but looking great so far 😄

**Comment by derickl** [comment]
I have access to Postgres 12 can confirm [both queries](https://github.com/sqlfluff/sqlfluff/pull/2041#issuecomment-986293871) are valid

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (derickl) [reply src/sqlfluff/dialects/dialect_postgres.py:1391]: I decided to add the complete syntax. I'll add some examples/tests tomorrow

Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_postgres.py:1391]: Amazing stuff, thanks! Let me know when they're added and then I can take a look 😄

**Comment by derickl** [comment]
@jpy-git I believe this should now be ready for your review. I've been unable to figure out how to support 'WITH [NO] DATA' without parenthesis

**Comment by jpy-git** [comment]
Awesome thanks, I should have time to review later this evening

File: test/fixtures/dialects/postgres/postgres_drop_materialized_view.sql
Comment (jpy-git) [inline test/fixtures/dialects/postgres/postgres_drop_materialized_view.sql:6]: need unit test for `DROP MATERIALIZED VIEW foo, bar;`

File: test/fixtures/dialects/postgres/postgres_create_materialized_view.sql
Comment (jpy-git) [inline test/fixtures/dialects/postgres/postgres_create_materialized_view.sql:1]: Need an IF NOT EXISTS unit test for
```sql
CREATE MATERIALIZED VIEW IF NOT EXISTS my_mat_view AS
SELECT foo
FROM bar;
```

File: src/sqlfluff/dialects/dialect_postgres_keywords.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres_keywords.py:160]: This is fine here

File: src/sqlfluff/dialects/dialect_postgres_keywords.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres_keywords.py:301]: This list is made to match the official docs (https://www.postgresql.org/docs/14/sql-keywords-appendix.html). Could you move this keyword down to the `postgres_nondocs_keywords` section below?

File: src/sqlfluff/dialects/dialect_postgres_keywords.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres_keywords.py:451]: This list is made to match the official docs (https://www.postgresql.org/docs/14/sql-keywords-appendix.html). Could you move this keyword down to the `postgres_nondocs_keywords` section below?

File: src/sqlfluff/dialects/dialect_postgres_keywords.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres_keywords.py:577]: This list is made to match the official docs (https://www.postgresql.org/docs/14/sql-keywords-appendix.html). Could you move this keyword down to the `postgres_nondocs_keywords` section below?

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1415]:
```suggestion
                    Sequence("SET", "SCHEMA", Ref("SchemaReferenceSegment")),
```

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1473]:
```suggestion
                            Sequence(
                                Ref("ParameterNameSegment"),
                                Ref("EqualsSegment"),
                                Ref("LiteralGrammar"),
                            ),
```
value is not optional here
image

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_postgres.py:1508]: where does this come from, I couldn't find it in the docs? https://www.postgresql.org/docs/14/sql-altermaterializedview.html

**Review by jpy-git — COMMENTED** [review]
A couple things to look into, but otherwise looking good 👍

File: src/sqlfluff/dialects/dialect_postgres.py
Comment (derickl) [reply src/sqlfluff/dialects/dialect_postgres.py:1505]: Nice catch. Fixing shortly

File: src/sqlfluff/dialects/dialect_postgres_keywords.py
Comment (derickl) [reply src/sqlfluff/dialects/dialect_postgres_keywords.py:301]: Thanks for this. I now know where to reference what goes where!

File: test/fixtures/dialects/postgres/postgres_create_materialized_view.sql
Comment (derickl) [reply test/fixtures/dialects/postgres/postgres_create_materialized_view.sql:1]: Done

File: test/fixtures/dialects/postgres/postgres_drop_materialized_view.sql
Comment (derickl) [reply test/fixtures/dialects/postgres/postgres_drop_materialized_view.sql:11]: Done

**Comment by derickl** [comment]
Ready for another review @jpy-git.

**Review by jpy-git — APPROVED** [review]
Looks great to me! ⭐ Just going to tag the other maintainers to see if they have a solution for the brackets vs no brackets issue before I merge 👍

**Comment by jpy-git** [comment]
@tunetheweb @WittierDinosaur this PR is ready to go other than an issue with brackets vs no-brackets described [here](https://github.com/sqlfluff/sqlfluff/pull/2041#issuecomment-986293871) (also affects CREATE TABLE AS). Are you aware of a way to fix this?

**Comment by WittierDinosaur** [comment]
You probably need to add WITH to the list of terminators in SelectableGrammar. That's just off the top of my head, not got the code in front of me right now

**Comment by jpy-git** [comment]
Having a look through the codebase I can see theres a similar case for `WITH NO SCHEMA BINDING` for the [Redshift CREATE VIEW](https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_VIEW.html) (although its implemented in dialect_ansi.py) and that works fine with and without brackets so I'll look at that.

**Comment by jpy-git** [comment]
@derickl I have created a [PR](https://github.com/derickl/sqlfluff/pull/1) to fix the WITH [NO] DATA issue (also fixes for CREATE TABLE AS). Can you please merge it into your branch and then we can get this PR merged in, thanks 😄

File: src/sqlfluff/dialects/dialect_ansi.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_ansi.py:3047]:
```suggestion
    type = "with_data_clause"
```
We can fix that type I missed here

**Comment by jpy-git** [comment]
@derickl because I've fixed the WithDataClause type here you'll just need to re-generate the test yml with `tox -e generate-fixture-yml` and then push those updated files and then we're good to go 👍

**Comment by derickl** [comment]
Changes pushed

**Comment by jpy-git** [comment]
@derickl Great work on this! Merged it in now 🚀✨

**Pull request merged**
